### DISCUSSION TITLE
Use variables from data files in js asset files

### DISCUSSION THREAD

**Discussion by madsem** [11ty/eleventy#2263] [Q&A]
I'm in a situation where I have to load a bunch of third party ja code (analytics etc).
That slowed the page down considerably.

Now I want to lazy load all that stuff, something like:

#main.js
```js
// Load tracking code lazily, so it's non-blocking.
const pixels = () => {

  const pixel_ids = {
    google_analytics_id: "{{ settings.google_analytics_id }}",
  };
  
  import("./modules/google-analytics").then((analytics) => analytics.init());

  // Initate all other code paths here...
};

pixels();
```

The problem is now, google analytics and also others, require an id.
This id is part of a settings file (some of it also in the netlify cms, so users can add those themselves).

Is it possible to somehow tell 11ty to scan the `main.js` and replace those placeholders in `pixel_ids`?
Or is that only possible writing a custom 11ty transform?

**Comment by cfjedimaster** [comment]
Remember that Eleventy templates can output non-HTML files. So you can have analytics.liquid output to analytics.js via the `permalink` front matter. You could use variables in there that will become static in the final JS file.

**Reply by madsem** [reply]
Hey @cfjedimaster , that is true, but then I'd lose the ability to use imports, editor syntax highlighting and so forth.
Especially when there are many of those third-party scripts, it would get quite messy imho. :D

**Reply by madsem** [reply]
PSPS: also another very big drawback would be: This would then not go through my webpack build... I.e: It would not get transpiled.

**Reply by madsem** [reply]
I was already thinking of just adding a `<script>` tag to my `base.njk` where I would create a config object.
But it is kind of an ugly solution imo and then config and actual code would be completely separate also.

Hoping there is a solution that will allow me to have 11ty replace that directly in the `js` files. :)

**Reply by cfjedimaster** [reply]
I wouldn't think config being separate would be ugly, but it would be another network call (load js, load config json). Maybe consider the beforeBuild Eleventy event? Read in your JS file, edit it, and save it back.

**Reply by madsem** [reply]
Hmmmm, yes I think you're right. Why would it be ugly. :)
I thought about it, and it honestly seems like the most forward way.

Add a config object in my `base.njk`, before the main.js loads. Then I can just reference those ID variables. 👍

**Reply by cfjedimaster** [reply]
If I get some time this weekend, I may blog a demo of this. I kinda want to see it.
